File: .github/build_compact_search_index.mjs
#!/usr/bin/env node

import fs from "node:fs";
import vm from "node:vm";

const [, , inputPath, outputPath] = process.argv;

if (!inputPath || !outputPath) {
    console.error("Usage: build_compact_search_index.mjs <searchindex.js> <output.json>");
    process.exit(1);
}

const context = { window: { search: {} } };
vm.runInNewContext(fs.readFileSync(inputPath, "utf8"), context, { filename: inputPath });

const source = context.window.search;
if (!source.index || !source.doc_urls) {
    throw new Error(`${inputPath} does not contain an mdBook search index`);
}

const index = source.index;
const storedDocs = index.documentStore.docs;
const storedInfo = index.documentStore.docInfo;
const documentCount = index.documentStore.length;
const documents = new Array(documentCount);

for (let ref = 0; ref < documentCount; ref++) {
    const doc = storedDocs[String(ref)];
    if (!doc) throw new Error(`Search document references must be contiguous; missing ${ref}`);
    documents[ref] = [doc.title || "", doc.body || "", doc.breadcrumbs || ""];
}

function encodeTypedArray(values, Type) {
    const typed = new Type(values);
    return Buffer.from(typed.buffer, typed.byteOffset, typed.byteLength).toString("base64");
}

function flattenTerms(root) {
    const entries = [];
    const stack = [["", root]];

    while (stack.length) {
        const [prefix, node] = stack.pop();
        if (node.df > 0) entries.push([prefix, node.docs]);

        for (const key of Object.keys(node)) {
            if (key !== "docs" && key !== "df") stack.push([prefix + key, node[key]]);
        }
    }

    entries.sort((a, b) => a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0);
    return entries;
}

const fields = index.fields.map((name) => {
    const entries = flattenTerms(index.index[name].root);
    const postingOffsets = new Uint32Array(entries.length + 1);
    let postingCount = 0;

    for (let i = 0; i < entries.length; i++) {
        postingOffsets[i] = postingCount;
        postingCount += Object.keys(entries[i][1]).length;
    }
    postingOffsets[entries.length] = postingCount;

    const postingDocs = new Uint32Array(postingCount);
    const postingTermFrequencies = new Float64Array(postingCount);
    let cursor = 0;

    for (const [, postings] of entries) {
        for (const [ref, value] of Object.entries(postings)) {
            postingDocs[cursor] = Number(ref);
            postingTermFrequencies[cursor] = value.tf;
            cursor++;
        }
    }

    const fieldLengths = new Uint32Array(documentCount);
    for (let ref = 0; ref < documentCount; ref++) {
        fieldLengths[ref] = storedInfo[String(ref)]?.[name] || 0;
    }

    return {
        name,
        terms: entries.map(([term]) => term),
        posting_offsets: encodeTypedArray(postingOffsets, Uint32Array),
        posting_docs: encodeTypedArray(postingDocs, Uint32Array),
        posting_term_frequencies: encodeTypedArray(postingTermFrequencies, Float64Array),
        field_lengths: encodeTypedArray(fieldLengths, Uint32Array),
    };
});

const compact = {
    version: 1,
    pipeline: index.pipeline,
    results_options: source.results_options,
    search_options: source.search_options,
    doc_urls: source.doc_urls,
    documents,
    fields,
};

fs.writeFileSync(outputPath, JSON.stringify(compact));
console.log(`Compact search index: ${inputPath} -> ${outputPath} (${documents.length} documents)`);

File: .github/build_compact_search_index.py
#!/usr/bin/env python3

import ast
import base64
import json
import sys
from array import array
from pathlib import Path


if len(sys.argv) != 3:
    raise SystemExit("Usage: build_compact_search_index.py <searchindex.js> <output.json>")

input_path = Path(sys.argv[1])
output_path = Path(sys.argv[2])
source_text = input_path.read_text(encoding="utf-8")
prefix = "window.search = Object.assign(window.search, JSON.parse('"
suffix = "'));"

if not source_text.startswith(prefix) or not source_text.rstrip().endswith(suffix):
    raise ValueError(f"{input_path} does not have the expected mdBook search-index wrapper")

encoded_json = source_text[len(prefix):source_text.rfind(suffix)]
source = json.loads(ast.literal_eval("'" + encoded_json + "'"))

if "index" not in source or "doc_urls" not in source:
    raise ValueError(f"{input_path} does not contain an mdBook search index")

index = source["index"]
stored_docs = index["documentStore"]["docs"]
stored_info = index["documentStore"]["docInfo"]
document_count = index["documentStore"]["length"]
documents = []

for ref in range(document_count):
    doc = stored_docs.get(str(ref))
    if doc is None:
        raise ValueError(f"Search document references must be contiguous; missing {ref}")
    documents.append([doc.get("title", ""), doc.get("body", ""), doc.get("breadcrumbs", "")])


def encode_array(values):
    if sys.byteorder != "little":
        values.byteswap()
    return base64.b64encode(values.tobytes()).decode("ascii")


def flatten_terms(root):
    entries = []
    stack = [("", root)]
    while stack:
        term, node = stack.pop()
        if node["df"] > 0:
            entries.append((term, node["docs"]))
        for key, child in node.items():
            if key not in ("docs", "df"):
                stack.append((term + key, child))
    entries.sort(key=lambda entry: entry[0])
    return entries


fields = []
for name in index["fields"]:
    entries = flatten_terms(index["index"][name]["root"])
    posting_offsets = array("I", [0])
    posting_docs = array("I")
    posting_term_frequencies = array("d")

    for _, postings in entries:
        # JavaScript enumerates integer-like object keys numerically.
        for ref in sorted(postings, key=int):
            value = postings[ref]
            posting_docs.append(int(ref))
            posting_term_frequencies.append(value["tf"])
        posting_offsets.append(len(posting_docs))

    field_lengths = array(
        "I",
        (stored_info.get(str(ref), {}).get(name, 0) for ref in range(document_count)),
    )
    if posting_offsets.itemsize != 4 or posting_docs.itemsize != 4 or field_lengths.itemsize != 4:
        raise RuntimeError("Compact index requires 32-bit unsigned integer arrays")
    if posting_term_frequencies.itemsize != 8:
        raise RuntimeError("Compact index requires 64-bit floating-point arrays")

    fields.append(
        {
            "name": name,
            "terms": [term for term, _ in entries],
            "posting_offsets": encode_array(posting_offsets),
            "posting_docs": encode_array(posting_docs),
            "posting_term_frequencies": encode_array(posting_term_frequencies),
            "field_lengths": encode_array(field_lengths),
        }
    )

compact = {
    "version": 1,
    "pipeline": index["pipeline"],
    "results_options": source.get("results_options"),
    "search_options": source.get("search_options"),
    "doc_urls": source["doc_urls"],
    "documents": documents,
    "fields": fields,
}

with output_path.open("w", encoding="utf-8") as output:
    json.dump(compact, output, ensure_ascii=False, separators=(",", ":"))

print(f"Compact search index: {input_path} -> {output_path} ({document_count} documents)")
